### PR TITLE
Camera Server & Camera Input Form

### PR DESCRIPTION
**Pull request by SeijDeLeon**
-Submit **python-server/pvCamera.py** for review
-Modified Camera react component to optionally accept user inputs for making camera ws connection
-Added a separate bridge docker file (for Mac os) so the default docker uses host mode

### CODE REVIEW

File: python-server/pvCamera.py [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+            rawImageArray, timestamp, updatedSettings = await buffer.get()
+
+            if updatedSettings:
+                currentSettings = updatedSettings
+                await websocket.send_text(json.dumps(currentSettings))
+                continue
+
+            height, width, colorMode, dataType = currentSettings['y'], currentSettings['x'], currentSettings['colorMode'], currentSettings['dataType']
+
+            try:
```
Comment (dylanmcreynolds) [inline python-server/pvCamera.py:175]: I would wrap array manipulation in its own non-async function then call it here with `asyncio.to_thread`. Numpy functions aren't async, but doing this allows multiple clients to run concurrently without being blocked.

File: python-server/pvCamera.py [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+        mode = 'RGB'
+    else:
+        raise ValueError(f"Unsupported color mode: {colorMode}")
+    
+    return reshaped_data, mode
+
+async def send_image(array_data, websocket, mode):
+    max_dimension = 2500 #maximum pixel width or height to be sent out. Increase if higher fidelity is needed
+    try:
+        if array_data.shape[0] > max_dimension or array_data.shape[1] > max_dimension:
```
Comment (dylanmcreynolds) [inline python-server/pvCamera.py:222]: You don't want to be running all of this buffer and image manipulation in an `async` function. I would:

- remove `async` keyword from this function
- pull the `await websocket.send_bytes(buffered.getvalue())` call out of this function
- call this function with `asyncio.run_thread()`

File: python-server/test/test_Camera.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
+import pytest
```
Comment (dylanmcreynolds) [inline python-server/test/test_Camera.py:1]: Minor comment that people module names should be snake_case (test_camera.py instead of test_Camera.py)

**Review by dylanmcreynolds — CHANGES_REQUESTED** [review]
Asyncio contains a foot-gun which is that anything you do inside an `async` function that is not awaitable will take the entire event-loop's attention. For a server, this means that requests will be blocked until they complete. I added a few comments about places to look at.

File: python-server/pvCamera.py [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+                #array_data = normalize_array_data(array_data, dataType)
+                #array_data, mode = reshape_array(array_data, height, width, colorMode)
+                #await send_image(array_data, websocket, mode)
+
+                #change to make non blocking, PR#13
+                format_array_coroutine = asyncio.to_thread(format_array, rawImageArray, height, width, colorMode, dataType)
+                array_data, mode = await format_array_coroutine
+                
+                get_buffer_coroutine = asyncio.to_thread(send_image, array_data, websocket, mode)
+                buffered = await get_buffer_coroutine
```
Comment (dylanmcreynolds) [inline python-server/pvCamera.py:192]: It's typical to await in the same line as defining the co-routine:

```

                #change to make non blocking, PR#13
                array_data, mode = await asyncio.to_thread(format_array, rawImageArray, height, width, colorMode, dataType)
                buffered = await asyncio.to_thread(send_image, array_data, websocket, mode)
           
```

Also, thinking more, let's combine these two separate threaded functions into one. Threads have a small cost, and reducing them is good when you can.

Comment (SeijDeLeon) [reply python-server/pvCamera.py]: -Put these two calls into new function `get_buffer()` line 217
-Called function same line as await
-Revised try catch blocks

File: python-server/test/test_Camera.py (hunk 3 above)
Comment (SeijDeLeon) [reply python-server/test/test_Camera.py]: -Revised to test_camera.py

**Pull request merged**